### PR TITLE
[Utility] Utility Spec Overhaul & Gateway Actor Introduction

### PR DESCRIPTION
**Pull request by Olshansk**
This PR is an iteration of the existing specifications that aims to make the spec more accessible to others, easier to update/maintain, and introduce a few new concepts to the protocol.

_Notes to the reviewers_:
- This PR does update sections 5 & 6 in detail
- This PR does not contain new concepts related to the governance protocol
- This PR is not meant to be _"perfect"_, but rather an iteration that makes it easier to read & maintain relative to the [previous version](https://github.com/pokt-network/pocket-network-protocol/blob/0bf1c03deaaacb53027cffee94beb6bf3e5d6efa/utility/README.md).
- The diff can be viewed under `utility/README.md` [here](https://github.com/pokt-network/pocket-network-protocol/pull/25/files) but a more accessible markdown rendered version of the file can be seen [here](https://github.com/pokt-network/pocket-network-protocol/blob/issues/23/aat/utility/README.md)

## Formatting Changes
- Added a table of contents
- Reducing the amount of large text blocks -> easier to read
- Convert diagrams to use `mermaid` -> easier to maintain
- Creating smaller sections -> easier to update & navigate

## General Changes
- Added context on requirements and state of the documents
- Added an `RPC Trilemma` visual

## Fisherman / Application
- Provided additional details on `Probabilistic Volume Computation`
- Remove the need for `NTP (Network Time Protocol)`
- Slightly modified the `commit-and-reveal mechanism` for TestScores
- Use `Ring Signatures` in place of the App/Fish handshake to enable incognito sampling

### Servicer / Application
- Added rate limiting using `Token Bucket` algorithm
- Added example of `GeoZones algorithms`

### Gateway / Application
- Added base specification for the `Gateway Actor`

### CODE REVIEW

**Comment by Olshansk** [comment]
@jacklaing @jessicadaugherty @mokn This is a significant change, so please see the description (changes & notes in the PR description) before reviewing.

I also invite others who have looked and thought about the specification in depth in the past (cc @moatus @DragonDmoney). This update will likely make it easier to build upon to incorporate the issues/suggestions you have brought up in the past.

File: utility/README.md
Comment (deblasis) [inline utility/README.md:1093]: I would consider also an optional short text field used to reference the reason why the action was taken. Could be a URL (IPFS or Forum come to mind)

File: utility/README.md
Comment (deblasis) [inline utility/README.md:1106]: I would consider also an optional short text field used to reference the reason why the action was taken. Could be a URL (IPFS or Forum come to mind)

File: utility/README.md
Comment (deblasis) [inline utility/README.md:1380]: I have left some comments related to this 💙 Maybe we could use a uint64 where each number corresponds to some primary key on an off-chain database something simple maybe like

| MemoID      | URL |
| ----------- | ----------- |
| 1      | ipfs://...       |
| 2   | ipfs://...        |

**Review by deblasis — CHANGES_REQUESTED** [review]
Some minor changes and a few questions/thinking-out-loud here and there but, man, this is great work! 🚀

File: utility/README.md
Comment (jessicadaugherty) [inline utility/README.md:929]: If StakePerAppDelegation value changes, do the App stakes get re-distributed per that new value, or are they grandfathered into the old value and only net new delegating applications take the new value?

File: utility/README.md
Comment (jessicadaugherty) [reply utility/README.md:954]: @deblasis uPOKT is the smallest denomination of POKT (10^0). Micro POKT :)

See all POKT denominations here: https://docs.pokt.network/learn/economics/token/

File: utility/README.md
Comment (jessicadaugherty) [reply utility/README.md:1129]: No testing; only develop ;)

**Review by jessicadaugherty — APPROVED** [review]
Overall great iteration of the spec. Left a few comments with suggested edits and clarification points. Do you have a high level timeline of when we need to revisit ideas.md?

**Comment by Olshansk** [comment]
Small updates: going to hold off on @jacklaing's feedback/review, and will then tend all the comments at once.

**Comment by PoktBlade** [comment]
https://docs.google.com/document/d/1z5vKqt0ARq6wovjiHmG2vPAhX4ODLNd_FexcP_7TODY/edit

Posted some of my questions / comments here since I couldn't comment.

**Comment by PoktBlade** [comment]
https://docs.google.com/document/d/1z5vKqt0ARq6wovjiHmG2vPAhX4ODLNd_FexcP_7TODY/edit leaving this doc for anyone to read with my comments/questions. Will go back and comment on the actual spec since it seems i can actually comment on this account

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:833]: Comment: Missing details on how delegation “state” will be stored on the persistent layer. (Maybe I am just missing something, but the entire persistent layer does not document the schema specifications). Why do we define [account and pool](https://github.com/pokt-network/pocket-network-protocol/tree/issues/23/aat/utility#371-account-structure), but not any other actor protos/structs in this spec?

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:853]: Suggestion: If applications can delegate to multiple gateways, then there should be a mechanism to control how many “relays” a specific gateway can serve to prevent any gateway exhaustion. Imagine a scenario where “higher volumed” gateways uses all of the applications relays per session. This then could be a nuisance as other gateways would be unable to serve a request despite having N applications delegated to them. The check could be done on the servicer level as part of the relay validation and adding an allotment as part of the delegation method (with parameters of course)

Another note: A normal application could “count” the relays sent on the client layer and compare it to relaysPerSession, whereas gateways would have no idea if they reached their delegated application relay limits without parsing the servicer errors. So gateways would have to optimistically send a relay without a guaranteed success. However, with an allotment per delegation, gateways would also be able to count.

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:861]: What is on-chain StartSession request, is this referring to the dispatch RPC request or an actual transaction? It was inferred that sessions are implicitly “started” / generated based off a hash of variables as seen below:

“The seed data for a session is an implementation detail that could be composed of multiple variables. It includes, but is not limited to, attributes such as LatestBlockHash, Timestamp, etc…”

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:923]: somewhat related to economic incentives, but we should consider refractoring the entire the [“Block rewards](https://github.com/pokt-network/pocket-network-protocol/tree/issues/23/aat/utility#362-block-rewards) to apply for all actors in the network in a generic form, so that way it is not inclusive to just DAO rewards and becomes a generic formula based off governance parameters. Example use case: Incentivizing gateway actors to run gateways and earn a section of rewards they help serve. After all, gateway actors have to “bond” some of their POKT through security deposits so i can expect some on chain incentivized reward to accompany it. I guess it can always come later, with the cost of a chain upgrade coordination.

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:757]: How are they handled through gateway model? Didn’t see any details about whether the gateway security deposit is burned or if it’s burned through the app. I assume the app.

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:934]: What’s the purpose of security deposit - is it to signify trust or later create economic incentives for running gateways?

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:932]: For gateways, did not find specifications about gateway “unregistering”, what happens to security deposits, governance parameters (bounded by a registering period?)..

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:831]: Missing specification around removing delegation (i.e if a gateway goes down)

File: utility/README.md
Comment (h5law) [inline utility/README.md:1250]: `It is the primary reason that Fisherman (at this stage) require DAO Approval.`
Sounds better here

File: utility/README.md
Comment (h5law) [inline utility/README.md:1272]: `It only reduces overall network inflation`
sounds better here

**Comment by Olshansk** [comment]
@deblasis @PoktBlade @jessicadaugherty Appreciate the review, questions and comments.

I resolved the "easy ones" and left comments/replies on the more difficult questions, sometimes updating the spec and sometimes updating the `ideas` section. PTAL and `Resolve` the comments where you're content with the answer or lmk if I should make further updates/changes/clarifications.

If all looks good, let's merge it in :)

File: utility/README.md
Comment (deblasis) [reply utility/README.md:274]: Yes it makes sense. 🙏

File: utility/README.md
Comment (deblasis) [reply utility/README.md]: Yes I think it's clearer if you changed as described

File: utility/README.md
Comment (PoktBlade) [reply utility/README.md:848]: For consistency purposes, I'd remove them besides the `SendTransaction` piece, but otherwise LGTM

File: utility/README.md
Comment (PoktBlade) [inline utility/README.md:298]: "andy"? -> "any"?

File: utility/README.md
Comment (jessicadaugherty) [reply utility/README.md:944]: Perfect!

**Review by jessicadaugherty — APPROVED** [review]
LFG

File: utility/README.md
Comment (PoktBlade) [reply utility/README.md:298]: What is _real_ geographic distribution? Given that there exists a `ServiceUrl` that eventually resolves into an IP address, this goes against [AnyCast](https://en.wikipedia.org/wiki/Anycast). Why do we seek to limit it?

Interesting change for when regenesis happens from V0 to V1. Limiting geographic zones definitely is more simplistic, however in practicality, probably not. Node runners will converge where the economic incentives are. Pocket V1 allows for lightweight servicers, so what you end up creating is an incentive to hop to the highest traffic region by transferring keys from one server to another, effectively creating a regional hotspot now and then.

I'll need to think about further implications, but definitely, one of those tops of mind is arbitrarily creating regional hotspots and making it hard to bootstrap more PoPs like a "central" / new geo zone without having more Pocket at stake.

**Review by PoktBlade — APPROVED** [review]

File: utility/README.md
Comment (Olshansk) [reply utility/README.md]: > What is real geographic distribution?

You are correct that there is no way to enforce a real physical location (tor, VPN, etc...), but the best we can do (and what crypto is about) is incentivize it.

> Why do we seek to limit it?

I see it as incentivizing literal physical geo-distribution of servers to have precedence of being matched with applications in some region. If a certain country starts banning web3 access, the physical location of a server could play an important role in decentralization.

> I'll need to think about further implications, but definitely, one of those tops of mind is arbitrarily creating regional hotspots and making it hard to bootstrap more PoPs like a "central" / new geo zone without having more Pocket at stake.

Agreed that there's a lot to think through an unpack here, and free-market economics could lead to interesting solutions. Happy to review a change to the spec in a future iteration when you have an interesting proposal.

**Pull request merged**
